### PR TITLE
Improved control over declared config vars

### LINKED ISSUE #234: Improve config parsing
`braid.base.api/register-config-var!` allows modules to declare optional or required config vars (read from mount-args).

https://github.com/braidchat/braid/blob/master/src/braid/base/api.cljc#L149

Ex. `braid.embeds-map` does: 
```(base/register-config-var! :google-maps-api-key)```

However:
  - there is no way to distinguish optional and required config vars
  - there is no way to know what config vars could be set
  - there is no way to know what the values of those config vars should be
  - if incorrect config vars are passed in, braid happily accepts them and subtle issues happen later

Part 1:
  - have `register-config-var!` take a second parameter, which is a [malli schema](https://github.com/metosin/malli/)
    ex. ```(base/register-config-var! :google-maps-api-key [:re #"some-regex-for-google-api-keys"])```
  - (may need to have a 3rd arg for optional vs required)
   - when 'reading' the config values (currently, [here](https://github.com/braidchat/braid/blob/master/src/braid/base/conf.clj#L20)), check the values against the registered schema
   - change all uses of `register-config-var!` throughout Braid to use the 2-arg form

Part 2:
 - provide a function in `braid.dev.core` to list the expected config

Below is some related code I pulled from another project:

```clojure
(defn parse [config schema]
  (if (malli/validate (malli.util/closed-schema schema) config)
    config
    (let [cause (malli/explain (malli.util/closed-schema schema) config)]
      (throw
       (ex-info (str "Config invalid\n"
                     (with-out-str
                       (pprint/pprint (malli.error/humanize cause)))) cause)))))
```

### PR DESCRIPTION
**Pull request by gashndolo**
Allow config vars to be either optional or required.

### CODE REVIEW

**Comment by jamesnvc** [comment]
Thanks Gershom! One thing off the bat here; you'll probably want to remove the Calva output file here; that was probably accidentally committed (and maybe add `/.calva` to the `.gitignore`)

**Comment by gashndolo** [comment]
Completely missed that.

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,4 @@ figwheel_server.log
 /datomic-pro-0.9.5697
 /lucene
 .cpcache
+/.calva
```
Comment (rafd) [inline .gitignore:22]: Add newline at end of file

File: src/braid/base/api.cljc [hunk 2]
```diff
@@ -145,12 +145,20 @@
...
+     
      (defn register-config-var!
        "Add a keyword to be read from `env` and added to the `config` state"
-       [k]
-       {:pre [(keyword? k)]}
-       (swap! braid.base.conf/config-vars conj k))
+         [k required-or-optional schema]
+       {:pre [(keyword? k)
+              (#{:required :optional} required-or-optional)
+              ;; TODO use malli internals to check schema
```
Comment (rafd) [inline src/braid/base/api.cljc:154]: Use function from malli to check this

**Comment by rafd** [comment]
Thanks @gashndolo. BTW the comments I've added are targeted at myself, before I merge in.

File: src/braid/base/conf.clj [hunk 3]
```diff
@@ -2,10 +2,43 @@
...
    [braid.core.hooks :as hooks]
+   [malli.core :as malli]
+   [malli.error :as malli.error]
+   [malli.transform :as malli.transform]
+   [clojure.pprint :as pprint]
    [mount.core :as mount :refer [defstate]]))
 
 (defonce config-vars
-  (hooks/register! (atom []) [keyword?]))
+  (hooks/register! (atom []) [any?]))
```
Comment (rafd) [inline src/braid/base/conf.clj:12]: Actually spec out instead of using `any?`

File: src/braid/chat/core.cljc [hunk 4]
```diff
@@ -31,23 +31,27 @@
...
+       (base/register-config-var! :github-client-secret :optional [:string]) 
+       (base/register-config-var! :hmac-secret :optional [:string]) 
+       (base/register-config-var! :http-only :optional [:string]) 
+       (base/register-config-var! :email-host :optional [:string]) 
+       (base/register-config-var! :email-user :optional [:string]) 
+       (base/register-config-var! :email-password :optional [:string]) 
+       (base/register-config-var! :email-port :optional [:string]) 
+       (base/register-config-var! :email-secure :optional [:string]) 
+       (base/register-config-var! :email-from :optional [:string]) 
+       (base/register-config-var! :site-url :optional [:string])
```
Comment (rafd) [inline src/braid/chat/core.cljc:52]: Some of these can be replaced with regular expressions as appropriate

**Pull request merged**